cargo/cargo_build_script.bzl

```diff
@@ -450,6 +450,17 @@ def _cargo_dep_env_implementation(ctx):
         outputs = [empty_dir],
         executable = "true",
     )
+
+    build_infos = []
+    if ctx.file.out_dir:
+        build_infos.append(BuildInfo(
```

Comment:
Why is this a separate `BuildInfo` provider? Is there a reason you couldn't just conditionally assign `empty_dir` or `ctx.file.out_dir`?

Reply:
Not sure which piece you're asking about, here's answers to two questions I can think of:
1. It seemed wasteful to create a `BuildInfo` even if there's no `out_dir` to manage.
2. A `BuildInfo` in `DepInfo.transitive_build_infos` is handled very differently than the main `BuildInfo` being returned, so they have to be separate. The former is is for build.rs scripts that depend on it transitively, and the latter is for direct dependencies of any kind.

Reply:
More directly, why does there need to be a new `BuildInfo` here vs updating the one created here:
https://github.com/bazelbuild/rules_rust/blob/0.10.0/cargo/cargo_build_script.bzl#L455-L462

Reply:
That's my #2: that `BuildInfo` is used in separate contexts from this new one, and I don't think `cargo_dep_env` should contribute settings there. Specifically, I don't see where `dep_env` in that `BuildInfo` is ever used, from what I see it just gets rolled up the dependency tree but nothing ever uses it. The `out_dir` from that `BuildInfo` does get used though, which I think would be pretty confusing to expose.

I think the `out_dir` from the main `BuildInfo` is intended for use in conjunction with `rustc_env`, which `cargo_dep_env` currently does not have the ability to set. I do not think adding that is necessary, because it's easy enough to write a `cargo_build_script` that passes values through. I'm not sure whether adding it would make sense, it gets kind of complicated to expose the dual meanings of `out_dir`, but at the same time I've got a `cargo_build_script` that just copies files from `dep_env` values to its `out_dir` and then sets `rustc_env` settings to point to them.

Reply:
Sorry for the delay here. I keep reading this and still don't think I've fully grokked the need for the transitive provider. The last two things I think I'll ask are:
1. Can you show me an example of something failing with `out_dir` being placed in the top level provider? Or is this more of a logical/correctness thing?
2. Can you add a code comment explaining the transitive nature here?

Reply:
1. Nothing fails, it just does something different. For example, my newly-added `//test/dep_env:build_read_dep_dir` fails because it can't find the file after changing that.
    
    The place this second (direct) BuildInfo would be useful is if a `cargo_dep_env` was directly depended on by a `rust_library` and set `rustc_env`.
2. Done